### PR TITLE
Develop to Main

### PR DESCRIPTION
**Pull request by YarinsHyper**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -43,88 +47,126 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+		"fileID",
+	)
+
+	return corsConfig
+}
+
 //CreateBirthday inserts a birthday object
 func (r *Router) CreateBirthday(c *gin.Context) {
-
+	var code = 201
```
Comment (Shahar-Y) [inline main.go]: no need for that here, you can just add the number in the `c.JSON`. Same for the rest of the code. 
Not mandatory for this PR.

File: main.go [hunk 2]
```diff
@@ -43,88 +47,126 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+}
+
+//GetAllBirthdays gets all birthday objects
+func (r *Router) GetAllBirthdays(c *gin.Context) {
+	var code = 200
+	request := &bpb.GetAllBirthdaysRequest{}
+	res, err := r.client.GetAllBirthdays(c, request)
+	if err != nil {
+		code = 404
+		fmt.Println("get all birthdays error.\n", err)
```
Comment (Shahar-Y) [inline main.go]: No `fmt.Println` in the code! It is the same as console.log. Remove from the rest of the code.
You may use a logger.

**Review by Shahar-Y — APPROVED** [review]
Approved. Notice not to push a PR without approval!
Good job :+1:

File: main.go (hunk 1 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 2 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go [hunk 3]
```diff
@@ -21,18 +22,20 @@ const (
...
-	if envError != nil {
-		log.Fatal("Error loading .env file! error: ", envError)
+	// Loading the dotenv parameters
+	config, err := util.LoadConfig(".")
+	if err != nil {
+		log.Fatal("cannot load config:", err)
 	}
 
-	address := os.Getenv("ADDRESS") + os.Getenv("PORT")
+	address := config.Address + config.Port
```
Comment (yoavrozner) [inline main.go]: Use one env for the address and port. See our api-gateway

**Review by yoavrozner — COMMENTED** [review]
Well done @yarinBenisty !
Please see my note

File: .env [hunk 4]
```diff
@@ -1,3 +0,0 @@
-PORT=8000
-ROUTER_PORT=9000
```
Comment (Shahar-Y) [inline .env:2]: Delete the env file completely

**Review by Shahar-Y — APPROVED** [review]

File: .env (hunk 4 above)
Comment (YarinsHyper) [reply .env:2]: done

File: main.go (hunk 3 above)
Comment (YarinsHyper) [reply main.go]: done

**Review by Shahar-Y — APPROVED** [review]

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+FROM golang:alpine
+# Set the Current Working Directory inside the container
+WORKDIR /app
+# Copy everything from the current directory to the PWD (Present Working Directory) inside the container
+COPY go.mod go.sum ./
+COPY app.env .
+# Download all the dependencies
+RUN go get -d -v ./...
```
Comment (nettashviro) [inline Dockerfile]: change to go mod download

File: main.go [hunk 6]
```diff
@@ -1,38 +1,44 @@
 package main
```
Comment (nettashviro) [inline main.go:1]: The structure of the folders and files in your service is incorrect, this logic shouldn't suppose to be in main.go

File: main.go [hunk 7]
```diff
@@ -1,38 +1,44 @@
...
 	bpb "github.com/yarinBenisty/birthday-service/proto"
 	"google.golang.org/grpc"
 )
 
 const (
-	name           = "name"
-	personalNumber = "personalNumber"
-	date           = "date"
+	// Name parameter
+	Name = "name"
```
Comment (nettashviro) [inline main.go]:
```suggestion
	ParamName = "name"
```

File: main.go [hunk 8]
```diff
@@ -1,38 +1,44 @@
...
+	// PersonalNumber parameter
+	PersonalNumber = "personalNumber"
```
Comment (nettashviro) [inline main.go]:
```suggestion
	ParamPersonalNumber = "personalNumber"
```

File: main.go [hunk 9]
```diff
@@ -1,38 +1,44 @@
...
+	// Date parameter
+	Date = "date"
```
Comment (nettashviro) [inline main.go]:
```suggestion
	ParamDate = "date"
```

File: main.go [hunk 10]
```diff
@@ -1,38 +1,44 @@
...
 
 // Func that creates a new client and connects
 // to the bd-service client through the proto file
 func initClientConnection() bpb.BirthdayFunctionsClient {
 
-	envError := godotenv.Load()
-	if envError != nil {
-		log.Fatal("Error loading .env file! error: ", envError)
+	// Loading the dotenv parameters
+	config, err := util.LoadConfig(".")
```
Comment (nettashviro) [inline main.go]: Why reload the config all over again?

File: main.go [hunk 11]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
 		log.Fatalln("connection error: ", err)
 		os.Exit(4)
```
Comment (nettashviro) [inline main.go]: remove this

File: main.go [hunk 12]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 	}
-	// defer conn.Close()
```
Comment (nettashviro) [inline main.go:46]: why did you delete this?

File: main.go [hunk 13]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+}
 
+// CreateBirthday inserts a birthday object if it
+// doesn't exist. if it does, its being overritten
+func (r *Router) CreateBirthday(c *gin.Context) {
 	request := &bpb.CreateBirthdayRequest{
-		PersonalNumber: c.Request.FormValue(personalNumber),
-		Name:           c.Request.FormValue(name),
-		Date:           c.Request.FormValue(date),
+		PersonalNumber: c.Request.FormValue(PersonalNumber),
```
Comment (nettashviro) [inline main.go]: please use c.bind instead and check if the request body is valid

File: main.go [hunk 14]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+		Name:           c.Request.FormValue(Name),
+		Date:           c.Request.FormValue(Date),
 	}
-
 	res, err := r.client.CreateBirthday(c, request)
 	if err != nil {
-		fmt.Println("create birthday error: ", err)
+		log.Fatal("create birthday error: ", err)
 		c.String(400, "create birthday method failed. \nerror: %s", err)
```
Comment (nettashviro) [inline main.go]: please use http.<status-code> from "net/http"

File: main.go [hunk 15]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-		Date:           c.Request.FormValue(date),
+		PersonalNumber: c.Request.FormValue(PersonalNumber),
+		Name:           c.Request.FormValue(Name),
+		Date:           c.Request.FormValue(Date),
 	}
-
 	res, err := r.client.CreateBirthday(c, request)
 	if err != nil {
-		fmt.Println("create birthday error: ", err)
+		log.Fatal("create birthday error: ", err)
```
Comment (nettashviro) [inline main.go]: it's not fatal

Comment (nettashviro) [reply main.go]: u must understand that once you call 'fatal', os.exit  is called

File: main.go [hunk 16]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 		c.String(400, "create birthday method failed. \nerror: %s", err)
-		os.Exit(3)
 	}
-	c.JSON(201, res)
+	c.JSON(200, res)
```
Comment (nettashviro) [inline main.go]: please use http.(statuscode) from "net/http"

File: main.go [hunk 17]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 
 	r := &Router{}
 	r.client = initClientConnection()
 
 	mainRouter := gin.Default()
+	mainRouter.Use(
+		cors.New(corsRouterConfig()),
+	)
+
 	mainRouter.POST("/api/createBirthday", r.CreateBirthday)
```
Comment (nettashviro) [inline main.go]: please change it to more restful way. 
```suggestion
	mainRouter.POST("/api/birthday", r.CreateBirthday)
```

File: main.go [hunk 18]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 	mainRouter.GET("/api/getBirthday", r.GetBirthday)
```
Comment (nettashviro) [inline main.go]:
```suggestion
	mainRouter.GET("/api/birthday/:userid", r.GetBirthday)
```

File: main.go [hunk 19]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-	mainRouter.POST("/api/updateBirthday", r.UpdateBirthday)
+	mainRouter.GET("/api/getAllBirthdays", r.GetAllBirthdays)
+	mainRouter.POST("/api/updateBirthday", r.CreateBirthday)
```
Comment (nettashviro) [inline main.go]:
```suggestion
	mainRouter.PUT("/api/birthday", r.CreateBirthday)
```

Comment (nettashviro) [reply main.go]: update and create is the same functions??

File: main.go [hunk 20]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 
 	mainRouter := gin.Default()
+	mainRouter.Use(
+		cors.New(corsRouterConfig()),
+	)
+
 	mainRouter.POST("/api/createBirthday", r.CreateBirthday)
 	mainRouter.GET("/api/getBirthday", r.GetBirthday)
-	mainRouter.POST("/api/updateBirthday", r.UpdateBirthday)
+	mainRouter.GET("/api/getAllBirthdays", r.GetAllBirthdays)
```
Comment (nettashviro) [inline main.go]:
```suggestion
	mainRouter.GET("/api/birthday", r.GetAllBirthdays)
```

File: main.go [hunk 21]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+	mainRouter.POST("/api/updateBirthday", r.CreateBirthday)
 	mainRouter.DELETE("/api/deleteBirthday", r.DeleteBirthday)
```
Comment (nettashviro) [inline main.go]:
```suggestion
	mainRouter.DELETE("/api/birthday", r.DeleteBirthday)
```

File: main.go [hunk 22]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 
-	err := mainRouter.Run(":" + routerPort)
+	err = mainRouter.Run(":" + routerPort)
```
Comment (nettashviro) [inline main.go:172]: please change the if err!=nil to the same line.

File: main.go [hunk 23]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-// DeleteBirthday deletes a certain birthday object by personal number
+// DeleteBirthday deletes a birthday object by personal number
 func (r *Router) DeleteBirthday(c *gin.Context) {
-
-	request := &bpb.DeleteBirthdayRequest{PersonalNumber: personalNumber}
+	request := &bpb.DeleteBirthdayRequest{PersonalNumber: c.Query(PersonalNumber)}
 	res, err := r.client.DeleteBirthday(c, request)
 	if err != nil {
-		fmt.Println("delete birthday error: ", err)
+		log.Fatal("delete birthday error: ", err)
```
Comment (nettashviro) [inline main.go]: same as my note before

File: main.go [hunk 24]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 		c.String(404, "delete birthday method failed. \nerror: %s", err)
```
Comment (nettashviro) [inline main.go]: please use http.(statuscode) from "net/http"

Comment (nettashviro) [reply main.go]: and it's not 404

File: main.go [hunk 25]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-		os.Exit(5)
 	}
 	c.JSON(204, res)
```
Comment (nettashviro) [inline main.go]: please use http.(statuscode) from "net/http"

File: main.go [hunk 26]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-	res, err := r.client.UpdateBirthday(c, request)
+//GetAllBirthdays returns all birthday objects
+func (r *Router) GetAllBirthdays(c *gin.Context) {
+	request := &bpb.GetAllBirthdaysRequest{}
+	res, err := r.client.GetAllBirthdays(c, request)
 	if err != nil {
-		fmt.Println("update birthday error: ", err)
-		c.String(400, "update birthday method failed. \nerror: %s", err)
-		os.Exit(5)
+		log.Fatal("get all birthdays error.\n", err)
```
Comment (nettashviro) [inline main.go]: same

File: main.go [hunk 27]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+		c.String(404, "get all birthday method failed. \nerror: %s", err)
```
Comment (nettashviro) [inline main.go]: http package

File: main.go [hunk 28]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
 	}
-	c.JSON(201, res)
+	c.JSON(200, res)
```
Comment (nettashviro) [inline main.go]: http package

File: main.go [hunk 29]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(personalNumber)}
+	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(PersonalNumber)}
 	res, err := r.client.GetBirthday(c, request)
 	if err != nil {
-		fmt.Println("get birthday error: ", err)
+		log.Fatal("get birthday error: ", err)
 		c.String(404, "get birthday method failed. \nerror: %s", err)
-		os.Exit(5)
 	}
 	c.JSON(200, res)
```
Comment (nettashviro) [inline main.go]: http package

File: main.go [hunk 30]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
+// GetBirthday returns a birthday object
 func (r *Router) GetBirthday(c *gin.Context) {
-
-	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(personalNumber)}
+	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(PersonalNumber)}
 	res, err := r.client.GetBirthday(c, request)
 	if err != nil {
-		fmt.Println("get birthday error: ", err)
+		log.Fatal("get birthday error: ", err)
 		c.String(404, "get birthday method failed. \nerror: %s", err)
```
Comment (nettashviro) [inline main.go]: Please separate the errors from not found birthday to different error  (server/unknown)

File: main.go [hunk 31]
```diff
@@ -43,88 +49,106 @@ func initClientConnection() bpb.BirthdayFunctionsClient {
...
-// GetBirthday read/get all birthday objects
+// GetBirthday returns a birthday object
 func (r *Router) GetBirthday(c *gin.Context) {
-
-	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(personalNumber)}
+	request := &bpb.GetBirthdayRequest{PersonalNumber: c.Query(PersonalNumber)}
 	res, err := r.client.GetBirthday(c, request)
 	if err != nil {
-		fmt.Println("get birthday error: ", err)
+		log.Fatal("get birthday error: ", err)
```
Comment (nettashviro) [inline main.go]: same

File: docker-compose.yml [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+    restart: always
+    ports:
+        - 27017:27017
+    environment:
+        MONGO_INITDB_ROOT_USERNAME: root
+        MONGO_INITDB_ROOT_PASSWORD: example
+    # volumes:
+    #     - "./data/db:/data/db"
+    
+  mongo-express:
```
Comment (nettashviro) [inline docker-compose.yml:13]: Why is this necessary?

**Review by nettashviro — CHANGES_REQUESTED** [review]
please see my comments

File: Dockerfile (hunk 5 above)
Comment (YarinsHyper) [reply Dockerfile]: done

File: main.go (hunk 7 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 8 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 9 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 11 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 12 above)
Comment (YarinsHyper) [reply main.go:46]: i made mongo client connection in a different function which means that if im defering inside it, at the end it'll return the client and just close the connection any ways.

File: main.go (hunk 14 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 15 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 16 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 17 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 18 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 19 above)
Comment (YarinsHyper) [reply main.go]: yes, i do upsert in birthday service

Comment (YarinsHyper) [reply main.go]: when its seperated(one for update, one for create) its duplicating code, sat about it with shahar

File: main.go (hunk 20 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 21 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 22 above)
Comment (YarinsHyper) [reply main.go:172]: done

File: main.go (hunk 23 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 24 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 25 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 26 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 27 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 28 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 29 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 30 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 31 above)
Comment (YarinsHyper) [reply main.go]: done

File: docker-compose.yml (hunk 32 above)
Comment (YarinsHyper) [reply docker-compose.yml:13]: done, used it for testing purposes

File: main.go (hunk 10 above)
Comment (YarinsHyper) [reply main.go]: when i tried , i couldn't load it once for the whole file.

Comment (YarinsHyper) [reply main.go]: and i use it inside two different functions

File: main.go (hunk 13 above)
Comment (YarinsHyper) [reply main.go]: done

**Comment by YarinsHyper** [comment]
i went through all your comments and did everything there was left to do (hopefully i didn't miss anything).

File: main.go (hunk 11 above)
Comment (nettashviro) [reply main.go]: You need to add fatal in this case. Because if you  can't connect to the birthday service it is critical. I meant here to delete the os.exit

File: main.go (hunk 13 above)
Comment (nettashviro) [reply main.go]: add the bind before the request

Comment (nettashviro) [reply main.go]: https://github.com/meateam/api-gateway/blob/feature/add-advanced-search/search/search.go

File: main.go (hunk 23 above)
Comment (nettashviro) [reply main.go]: why http.StatusConflict?

File: main.go (hunk 24 above)
Comment (nettashviro) [reply main.go]: why http.StatusConflict?

**Comment by nettashviro** [comment]
```
	if err != nil {
		httpStatusCode := gwruntime.HTTPStatusFromCode(status.Code(err))
		loggermiddleware.LogError(r.logger, c.AbortWithError(httpStatusCode, err))

		return
	}
```
please check that when you access the client you need to return this error

File: main.go (hunk 13 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 11 above)
Comment (YarinsHyper) [reply main.go]: done

File: main.go (hunk 24 above)
Comment (YarinsHyper) [reply main.go]: done

**Review by nettashviro — APPROVED** [review]
image
